#Author: Hisatake Ishida
#Date : 4/5/21
#Description: Computes phylogenyetic analysis usinf Neighbour Joining (NJ), 
#Unweighted Pair Group Method with Arithmetic mean (UPGMA), 
#Maximum likelihood method with bootstrap analysis on 
#imported multiple sequence alignment file. 

# Setting the working directory. 
setwd('')

# Installing the required packages. 
install.packages("readxl")
install.packages("ape")
install.packages("phangorn")
install.packages("seqinr")
install.packages("AnnotationBustR")

# Loading required packages. 
library(ape)
library(phangorn)
library(seqinr)
library(AnnotationBustR)
library("readxl")

# Importing multiple sequence alignment data generated by the Python script. 
philinoidea_alignment_raw <- read.dna("aligned_seq.fa", format="fasta")

# Converting FASTA file to “phyDat” object. 
philinoidea_phyDat <- phyDat(philinoidea_alignment_raw, type = "DNA", levels = NULL)

# Producing distance matrix based on the nucleotide evolution model. 
model_testing <- modelTest(philinoidea_phyDat)
print(model_testing)
DNA_dist_matrix <- dist.ml(philinoidea_phyDat, model="F81")

# Estimating tree by UPGMA and NJ. 
philinoidea_UPGMA <- upgma(DNA_dist_matrix)
philinoidea_NJ  <- NJ(DNA_dist_matrix)

# Plotting the estimated tree by UPGMA and NJ. 
plot(philinoidea_UPGMA, main="UPGMA")
plot(philinoidea_NJ, main = "Neighbor Joining")

# Calculating Parsimony score to examine two estimated trees (UPGMA and NJ)
parsimony(philinoidea_UPGMA,philinoidea_phyDat)
parsimony(philinoidea_NJ,philinoidea_phyDat)

# Bootstrap analysis of NJ tree.
set.seed(123)
NJ_bs <- bootstrap.phyDat(philinoidea_phyDat, FUN=function(x)NJ(dist.hamming(x)), bs=100)
plotBS(philinoidea_NJ, NJ_bs, "phylogram", p = 50, bs.col = 'blue')

# Constructing tree through Maximum Likelihood using NJ tree information.
maximum_likelihood_tree <- pml(philinoidea_NJ, philinoidea_phyDat)
print(maximum_likelihood_tree)

# Otimizing tree topology based on the evolution model. 
fit_maximum_likelihood_tree_GTR <- optim.pml(maximum_likelihood_tree, model = "GTR", rearrangement = "stochastic")
logLik(fit_maximum_likelihood_tree_GTR)

# Running Bootstrap analysis.
boot_strap <- bootstrap.pml(fit_maximum_likelihood_tree_GTR, bs=100, optNni=TRUE, multicore=TRUE, control = pml.control(trace=0))
plotBS(midpoint(fit_maximum_likelihood_tree_GTR$tree), bs.col = 'blue', boot_strap, p=40, "phylogram")
plotBS(midpoint(fit_maximum_likelihood_tree_GTR$tree), bs.col = 'blue', boot_strap, "phylogram")

# Exporting the final tree file.
write.tree(philinoidea_UPGMA, file="tree_UPGMA.tre")
write.tree(philinoidea_NJ, file="tree_NJ.tre")
write.tree(fit_maximum_likelihood_tree_GTR$tree, file="fit_maximum_likelihood_tree_GTR.tre")
